library/lua/gui/widgets.lua

```diff
@@ -2433,4 +2433,131 @@ function RangeSlider:onRenderBody(dc, rect)
     end
 end
 
+GraphicButton = defclass(GraphicButton, ResizingPanel)
+GraphicButton.ATTRS {
+    tiles = DEFAULT_NIL,
+    height = 0,
+    width = 0,
+    graphics = '',
+    graphic_file_dims = { x = 0, y = 0 },
+    pos = 0,
+    selected = false,
+    group_index = -1,
+    on_click_callback = DEFAULT_NIL
+}
+
+function GraphicButton:init()
+    self.tiles = {}
+    for y = 1, self.height do
+        for x = 1, self.width do
+            self.tiles[x] = {}
+            self.tiles[x][y] = nil
+        end
+    end
+end
+
+function GraphicButton:updateLayout(parent_rect)
+    self.subviews = {}
+    self:import(dfhack.textures['get' .. self.graphics .. 'TexposStart'](), self.pos,
+        self.graphic_file_dims.x, self.graphic_file_dims.y)
+
+    GraphicButton.super.updateLayout(self, parent_rect)
+end
+
+function GraphicButton:import(start, pos, dimx, dimy)
+    local offset = self.width * pos
+    local button_icon_start = start + offset
+    for y = 1, self.height do
+        local tiles = {}
+        for x = 1, self.width do
+            local tile = (button_icon_start + (x - 1)) + (y - 1) * dimx * self.width +
+                (self.selected and self.height * self.width * dimx or 0)
+            table.insert(tiles, { tile = tile })
+        end
+        self:addviews {
+            Label {
+                -- view_id = "button_but",
+                frame = { t = y, l = 0 },
+                text = tiles,
+                on_click = function()
+                    if self.on_click_callback then self.on_click_callback({ group_index = self.group_index }) end
+                    self.selected = not self.selected
+                    self.tiles = nil
+                    self:updateLayout()
```

Comment:
instead of re-initializing when the selected state changes, how about having two `Label`s with `visible` properties that are controlled by `self.selected`? that would allow all `import`ing to happen in `init`, with display state handled by existing widget logic